src/entry-client.js

```diff
@@ -32,35 +33,84 @@ export default function initializeClient(createApp, clientOpts) {
         // dynamic modules until Vuex implements a hasModule() type method.  See:
         //   https://github.com/vuejs/vuex/issues/833
         //   https://github.com/vuejs/vuex/pull/834
-        const registeredModules = {};
+        const registeredModules = [];
+
+        // Allow a function to be passed that can generate a route-aware
+        // module name
+        const getModuleName = (c, route) => (
+            typeof c.vuex.moduleName === 'function' ?
+                c.vuex.moduleName(route) :
+                c.vuex.moduleName
+        );
 
         // Before routing, register any dynamic Vuex modules for new components
         router.beforeResolve((to, from, next) => {
-            router.getMatchedComponents(to)
-                .filter(c => 'vuex' in c && !registeredModules[c.vuex.moduleName])
-                .forEach((c) => {
-                    opts.logger.info('Registering dynamic Vuex module:', c.vuex.moduleName);
-                    store.registerModule(c.vuex.moduleName, c.vuex.module, {
-                        preserveState: store.state[c.vuex.moduleName] != null,
+            try {
+                router.getMatchedComponents(to)
+                    .filter(c => 'vuex' in c)
+                    .forEach((c) => {
+                        const moduleName = getModuleName(c, to);
+
+                        // Short circuit if we already have this module
+                        const idx = registeredModules.indexOf(moduleName);
+                        if (idx >= 0) {
+                            // Bump this module back to the end of the queue
+                            registeredModules.splice(idx, 1);
+                            registeredModules.push(moduleName);
+                            opts.logger.info(
+                                'Skipping duplicate Vuex module registration:',
+                                moduleName,
+                            );
```

Comment:
We're maintaining a queue of registered modules, so if we find a dup module (back button or routing to a prior page), we just move that back to the end (most recent spot) of the queue, but we won't re-register, etc.